### PR TITLE
support save/load checkpoint of Go PS

### PR DESCRIPTION
**Pull request by QiJune**

### CODE REVIEW

File: elasticdl/pkg/ps/checkpoint.go [hunk 1]
```diff
@@ -112,10 +118,10 @@ func LoadModelFromCheckPoint(checkPointDir string, shardID int, shardNum int) (*
 	return model, nil
 }
 
-// SaveModelToCheckPoint saves in-memory model to checkpoint
-func SaveModelToCheckPoint(checkPointDir string, model *Model, shardID int, shardNum int) {
-	os.MkdirAll(checkPointDir, os.ModePerm)
+// SaveModelToCheckpoint saves in-memory model to checkpoint
+func SaveModelToCheckpoint(checkpointDir string, model *Model, shardID int, shardNum int) {
+	os.MkdirAll(checkpointDir, os.ModePerm)
```
Comment (workingloong) [inline elasticdl/pkg/ps/checkpoint.go:123]: We may need to check whether the `checkpointDir` exists.

File: elasticdl/pkg/ps/server.go [hunk 2]
```diff
@@ -63,23 +71,53 @@ func createMasterClient(masterAddr string) *MasterClient {
 }
 
 // NewServer creates a Server instance
-func NewServer(ID int, optType string, optArgs string, masterAddr string, evaluationSteps int) *Server {
+func NewServer(ID int, optType string, optArgs string, masterAddr string,
```
Comment (workingloong) [inline elasticdl/pkg/ps/server.go:74]: Rename `NewServer`, maybe `ServerFactory`?

File: elasticdl/pkg/ps/server.go [hunk 3]
```diff
@@ -63,23 +71,53 @@ func createMasterClient(masterAddr string) *MasterClient {
...
 	if err != nil {
 		log.Fatalf("failed to create PS server: %v", err)
 	}
 	ps.ID = ID
 	ps.masterClient = createMasterClient(masterAddr)
-	ps.evaluationSteps = int32(evaluationSteps)
+	ps.evaluationStep = evaluationStep
+	ps.checkpointDirForInit = checkpointDirForInit
+	ps.checkpointDir = checkpointDir
+	ps.checkpointStep = checkpointStep
```
Comment (workingloong) [inline elasticdl/pkg/ps/server.go:99]: Duplicated with line 96

File: elasticdl/pkg/ps/server.go [hunk 4]
```diff
@@ -63,23 +71,53 @@ func createMasterClient(masterAddr string) *MasterClient {
...
+
+func (s *Server) saveCheckpointIfNeeded(modelVersion int) {
+	if s.checkpointDir != "" && s.checkpointStep != 0 && modelVersion%s.checkpointStep == 0 {
+		checkpointVersionDir := path.Join(s.checkpointDir, fmt.Sprintf("version-%d", modelVersion))
+		s.savedCheckpointDirs = append(s.savedCheckpointDirs, checkpointVersionDir)
+		SaveModelToCheckpoint(checkpointVersionDir, s.Model, s.ID, s.numPsPods)
+		if len(s.savedCheckpointDirs) > s.keepCheckpointMax {
+			deletedDir := s.savedCheckpointDirs[0]
+			s.savedCheckpointDirs = s.savedCheckpointDirs[1:]
+			os.RemoveAll(deletedDir)
```
Comment (workingloong) [inline elasticdl/pkg/ps/server.go]: The `deleteDir` may have been deleted by other ps instance. We can only allow the first ps instance to delete the old checkpoint directories.

File: elasticdl/pkg/ps/checkpoint.go (hunk 1 above)
Comment (QiJune) [reply elasticdl/pkg/ps/checkpoint.go:123]: From the document, https://golang.org/pkg/os/#MkdirAll, "If path is already a directory, MkdirAll does nothing and returns nil."

File: elasticdl/pkg/ps/server.go (hunk 2 above)
Comment (QiJune) [reply elasticdl/pkg/ps/server.go:74]: It's the naming convention of Go, we usually use `NewStruct` to create a Struct instance.

File: elasticdl/pkg/ps/server.go (hunk 3 above)
Comment (QiJune) [reply elasticdl/pkg/ps/server.go:99]: line 96 is evaluationStep, this line is checkpointStep

File: elasticdl/pkg/ps/server.go (hunk 4 above)
Comment (QiJune) [reply elasticdl/pkg/ps/server.go]: Good idea. Fix it.

**Review by skydoorkai — APPROVED** [review]

**Pull request merged**